ng-dev/auth/shared/oauth.ts

```diff
@@ -0,0 +1,239 @@
+        device_code: string;
+      },
+  );
+
+  Log.info(`Please visit: ${response.verification_uri || response.verification_url}`);
+  Log.info(`Enter your one time ID code: ${response.user_code}`);
+
+  /**
+   * The number of milliseconds to add to the requested internal from Google, utilized if Google requests
+   * that our requests slow down.
+   */
+  let pollingBackoff = 2500;
+
+  const oauthDanceTimeout = Date.now() + response.expires_in * 1000;
+
+  while (true) {
+    if (Date.now() > oauthDanceTimeout) {
+      throw {
+        authenticated: false,
+        message: 'Failed to completed OAuth authentication before the user code expired.',
+      };
+    }
+    // Wait for the requested interval before polling, this is done before the request as it is unnecessary to
+    //immediately poll while the user has to perform the auth out of this flow.
+    await new Promise((resolve) => setTimeout(resolve, response.interval * 1000 + pollingBackoff));
+
+    const result = await pollAuthServer(
+      authConfig.tokenEndpoint,
+      response.device_code,
+      client_id,
+      client_secret,
+    );
+    if (!result.error) {
+      return {
+        ...result,
+        idToken: result.id_token,
+        accessToken: result.access_token,
+      };
+    }
+    if (result.error === 'access_denied') {
+      throw {
+        authenticated: false,
+        message: 'Unable to authorize, as access was denied during the OAuth flow.',
+      };
+    }
+
+    if (result.error === 'authorization_pending') {
+      // Update messaging.
+    }
+
+    if (result.error === 'slow_down') {
+      // Update messaging.
+      pollingBackoff += 5000;
+    }
+
+    if (
+      result.error === 'invalid_client' ||
+      result.error === 'unsupported_grant_type' ||
+      result.error === 'invalid_grant' ||
+      result.error === 'invalid_request'
+    ) {
+      throw {
+        authenticated: false,
+        message: result.errorDescription,
+      };
+    }
+  }
+}
+
+async function pollAuthServer(
+  serverUrl: string,
+  deviceCode: string,
+  clientId: string,
+  clientSecret?: string,
+) {
+  const url = new URL(serverUrl);
+  if (clientSecret) {
+    url.searchParams.append('client_secret', clientSecret);
+  }
+  url.searchParams.append('client_id', clientId);
+  url.searchParams.append('device_code', deviceCode);
+  url.searchParams.append('grant_type', 'urn:ietf:params:oauth:grant-type:device_code');
+
+  return await fetch(url.toString(), {
+    method: 'POST',
+    headers: {'Accept': 'application/json'},
+  }).then((x) => x.json() as Promise<any>);
+}
+
+export const GoogleOAuthDanceConfig: OAuthDanceConfig = {
+  loopback: {
+    client_id: '823469418460-ta0oojev0ovg2qlmv6kn46qiaebkr4gi.apps.googleusercontent.com',
+    client_secret: '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',
+  },
+  oob: {
+    client_id: '823469418460-puj3s53su005dp2ima1uim3gil2uggur.apps.googleusercontent.com',
+    client_secret: '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',
```

Comment:
Can you add some details here on where this information comes from? e.g. the client id. to which OAuth app does it map etc.

Also, do we want to store a secret here publicly? is a secret even needed? I read somewhere it might not be needed.

Reply:
Its okay to store it publicly, we don't need the secret for the server, but the tooling we are relying on requires it if I remember correctly.

Reply:
From the google oauth docs:

>The process results in a client ID and, in some cases, a client secret, which you embed in the source code of your application. (In this context, the client secret is obviously not treated as a secret.)

Seems like this is okay. Maybe also for the Github Oauth app? I think we should check and add the corresponding docs/notes to clarify this being intended.